compiler/optimizations/inferConstRefs.cpp

```diff
@@ -23,62 +23,109 @@
+// This file implements the function 'inferConstRefs', which attempts to
+// determine whether or not a symbol can be const-val, const-ref, and/or
+// ref-to-const.
+//
+// At the beginning of the pass, we build up a list of SymExprs for each
+// VarSymbol and ArgSymbol. Each helper function typically iterates over each
+// SymExpr once. This allows us to handle recursion in a re-entrant fashion,
+// where recursive calls to helper functions continue where the earlier call
+// left off. This list of SymExprs, along with other metadata about the Symbol,
+// is stored in a ConstInfo instance.
+//
+// While SymExprs are being processed, we will typically break out of the list
+// of SymExprs early if we encounter a case that violates the constness we're
+// looking for. Either way, once we're done processing the SymExprs the helper
+// function can change the Qualifier or flags on the Symbol if applicable. In
+// the case of recursion, only the first call on the Symbol can update the
+// Qualifier or flags. This avoids potential cases where the recursive calls
+// find nothing wrong (and might otherwise update flags), but the earlier call
+// finds that the Symbol cannot be const.
+//
+// Once we've made a determination either way, the helper function sets the
+// appropriate 'finalized*' field in its ConstInfo instance. If other Symbols
+// then query the constness of this Symbol, we can quickly determine the
+// answer. The helper functions also reset the list of SymExprs to the head, so
+// that later helper functions can use the list re-entrantly.
+//
+// Until we clarify constness more in the language, this pass will not consider
+// a record to be const if any of its fields are modified.
+//
+
+class ConstInfo {
   public:
-    bool                 finalized; // Indicates if analysis is done
-    Symbol*              sym;
-    std::deque<SymExpr*> todo;
-    RefInfo(Symbol*);
+    bool                  finalizedConstness;
+    bool                  finalizedRefToConst;
+    bool                  alreadyCalled;
+    Symbol*               sym;
+    SymExpr*              fnUses;
+
+    // TODO: Should we use the linked-list embedded in each Symbol?
+    std::vector<SymExpr*> todo;
+
+    ConstInfo(Symbol*);
+    SymExpr* next();
+    void reset();
+    bool hasMore();
+
+  private:
+    int curTodo;
 };
 
-RefInfo::RefInfo(Symbol* s) {
+ConstInfo::ConstInfo(Symbol* s) {
   sym = s;
-  finalized = sym->qualType().getQual() == QUAL_CONST_REF;
+  fnUses = NULL;
+  Qualifier q = sym->qualType().getQual();
+  finalizedConstness = q == QUAL_CONST_REF || q == QUAL_CONST_VAL;
+  if (ArgSymbol* arg = toArgSymbol(s)) {
+    // Work around a bug where we can have an arg with the in-intent, but it's
+    // a reference
+    // BHARSH TODO: this shouldn't be possible in the qualified refs impl.
+    if (arg->intent == INTENT_CONST_IN && arg->isRef()) {
+      finalizedConstness = false;
+    }
+  }
+  finalizedRefToConst = sym->hasFlag(FLAG_REF_TO_CONST);
```

Comment:
Yeah. I view it as connected to the const-language-design questions but it's not necessarily so.